File: AccuPay/Domain/TimeLogs/TimeLogsReader.vb
Option Strict On

Imports System.IO
Imports System.Text.RegularExpressions
Imports AccuPay.Entity
Imports AccuPay.Utilities
Imports AccuPay.Utilities.Extensions

Namespace Global.AccuPay.Helper.TimeLogsReader

    Public Class TimeLogsReader

        Public Const FILE_NOT_FOUND_ERROR As String =
            "Import file not found. It may have been deleted or moved."

        'currently not used after using ObjectUtils conversion
        Private Const DateTimePattern As String = "^(\d{4})[\/\-](\d{1,2})[\/\-](\d{1,2}) (\d{1,2}):(\d{1,2}):?(\d{1,2})?$"

        Public Function Import(filename As String) As ImportOutput
            Dim output As New ImportOutput

            Dim logs = New List(Of ImportTimeAttendanceLog)

            Dim lineNumber As Integer = 0
            Dim lineContent As String

            Try
                Using fileStream = New FileStream(filename, FileMode.Open, FileAccess.Read, FileShare.ReadWrite), stream = New StreamReader(fileStream)
                    Do
                        lineNumber += 1
                        lineContent = stream.ReadLine()

                        Dim log = ParseLine(lineContent, lineNumber)

                        If log Is Nothing Then
                            Continue Do
                        Else
                            logs.Add(log)
                        End If

                    Loop Until lineContent Is Nothing
                End Using
            Catch ex As FileNotFoundException
                output.IsImportSuccess = False
                output.ErrorMessage = FILE_NOT_FOUND_ERROR
            Catch ex As Exception
                output.IsImportSuccess = False
                output.ErrorMessage = "An error occured. Please try again or contact Globagility."
            End Try

            output.Logs = logs.Where(Function(l) l.HasError = False).ToList
            output.Errors = logs.Where(Function(l) l.HasError = True).ToList

            Return output
        End Function

        Private Function ParseLine(lineContent As String, lineNumber As Integer) As ImportTimeAttendanceLog
            Try
                If String.IsNullOrEmpty(lineContent) Then
                    'not considered error if it's just an empty line
                    'but it won't still be added to TimeAttendanceLog
                    Return Nothing
                End If

                Dim parts = Regex.Split(lineContent, "\t")

                If parts.Length < 2 Then
                    Return (New ImportTimeAttendanceLog() With {
                        .LineContent = lineContent,
                        .LineNumber = lineNumber,
                        .ErrorMessage = "Needs at least 2 items in one line separated by a tab."
                    })
                End If

                If String.IsNullOrEmpty(parts(0)) AndAlso String.IsNullOrEmpty(parts(1)) Then
                    'happens when blank lines was still read
                    'usually because it received an input like tab
                    Return Nothing
                End If

                Dim employeeNo = Trim(parts(0))

                Dim logDate = ObjectUtils.ToNullableDateTime(parts(1))

                If logDate Is Nothing Then
                    Return (New ImportTimeAttendanceLog() With {
                        .LineContent = lineContent,
                        .LineNumber = lineNumber,
                        .ErrorMessage = "Second column must be a valid Date Time."
                    })
                End If

                Return (
            New ImportTimeAttendanceLog() With {
                .EmployeeNumber = employeeNo,
                .DateTime = CType(logDate, Date),
                .LineContent = lineContent,
                .LineNumber = lineNumber
            })
            Catch ex As Exception
                Return (New ImportTimeAttendanceLog() With {
                        .LineContent = lineContent,
                        .LineNumber = lineNumber,
                        .ErrorMessage = "Error reading the line. Please check the template."
                })
            End Try
        End Function

        Public Class ImportOutput
            Public Property Logs As IList(Of ImportTimeAttendanceLog)
            Public Property Errors As IList(Of ImportTimeAttendanceLog)

            ''' <summary>
            ''' True if the file was read successfully. Even if there are errors parsing
            ''' some lines, as long as the file was read, this is still True.
            ''' Posible reason for this to become False is when it did not find the chosen
            ''' file.
            ''' </summary>
            Public Property IsImportSuccess As Boolean

            Public Property ErrorMessage As String

            Sub New()
                Me.Logs = New List(Of ImportTimeAttendanceLog)
                Me.Errors = New List(Of ImportTimeAttendanceLog)
                Me.IsImportSuccess = True

                Me.ErrorMessage = Nothing
            End Sub

        End Class

    End Class

    Public Class ImportTimeAttendanceLog

        ' - Contents and error message
        Public Property LineContent As String

        Public Property LineNumber As Integer
        Public Property ErrorMessage As String

        ' - Extracted contents
        Public Property EmployeeNumber As String

        Public Property [DateTime] As Date

        ' - analyzed data
        Public Property IsTimeIn As Boolean?

        Public Property LogDate As Date?
        Public Property Employee As Employee
        Public Property ShiftSchedule As ShiftSchedule
        Public Property EmployeeDutySchedule As EmployeeDutySchedule
        Public Property ShiftTimeInBounds As Date
        Public Property ShiftTimeOutBounds As Date
        Public Property WarningMessage As String

        Public Overrides Function ToString() As String
            Return DateTime.ToString("MM/dd/yyyy hh:mm tt")
        End Function

        Sub New()
            Me.IsTimeIn = Nothing
            Me.LogDate = Nothing

            Me.EmployeeNumber = Nothing

        End Sub

        Public ReadOnly Property Type As String
            Get
                Return If(IsTimeIn, "Time-in", "Time-out")
            End Get
        End Property

        Public ReadOnly Property ShiftDescription As String
            Get
                If ShiftSchedule IsNot Nothing AndAlso
                ShiftSchedule.Shift IsNot Nothing Then

                    Return $"{ShiftSchedule.Shift.TimeFrom.ToStringFormat("hh:mm tt")} - { _
                    ShiftSchedule.Shift.TimeTo.ToStringFormat("hh:mm tt")}"

                ElseIf EmployeeDutySchedule IsNot Nothing Then

                    Return $"{EmployeeDutySchedule.StartTime.ToStringFormat("hh:mm tt")} - { _
                    EmployeeDutySchedule.EndTime.ToStringFormat("hh:mm tt")}"

                End If

                Return "-"

            End Get
        End Property

        Public ReadOnly Property EmployeeFullName As String
            Get
                If Employee Is Nothing Then Return ""

                Return Employee.FirstName & " " &
                    If(String.IsNullOrWhiteSpace(Employee.MiddleName), "", Employee.MiddleName(0) + ". ") &
                    Employee.LastName
            End Get
        End Property

        Public ReadOnly Property HasError As Boolean
            Get
                Return ErrorMessage IsNot Nothing
            End Get
        End Property

        Public ReadOnly Property HasWarning As Boolean
            Get
                Return WarningMessage IsNot Nothing
            End Get
        End Property

        Public Shared Function GroupByEmployee(logs As IList(Of ImportTimeAttendanceLog)) _
            As List(Of IGrouping(Of String, ImportTimeAttendanceLog))

            Return logs.GroupBy(Function(l) l.EmployeeNumber).ToList()
        End Function

    End Class

End Namespace

File: Accupay.Data/Services/TimeLogParsing/TimeLogsReader.cs
using AccuPay.Data.Entities;
using AccuPay.Utilities;
using AccuPay.Utilities.Extensions;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

namespace AccuPay.Data.Services
{
    public class TimeLogsReader
    {
        public const string FILE_NOT_FOUND_ERROR = "Import file not found. It may have been deleted or moved.";

        public ImportOutput Import(string filename)
        {
            ImportOutput output = new ImportOutput();

            var logs = new List<ImportTimeAttendanceLog>();

            int lineNumber = 0;
            string lineContent;

            try
            {
                using (var fileStream = new FileStream(filename, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
                {
                    using (var stream = new StreamReader(fileStream))
                    {
                        do
                        {
                            lineNumber += 1;
                            lineContent = stream.ReadLine();
                            var log = ParseLine(lineContent, lineNumber);
                            if (log == null)
                                continue;
                            else
                                logs.Add(log);
                        }
                        while (lineContent != null);
                    }
                }
            }
            catch (FileNotFoundException)
            {
                output.IsImportSuccess = false;
                output.ErrorMessage = FILE_NOT_FOUND_ERROR;
            }
            catch (Exception)
            {
                output.IsImportSuccess = false;
                output.ErrorMessage = "An error occured. Please try again or contact Globagility.";
            }

            output.Logs = logs.Where(l => l.HasError == false).ToList();
            output.Errors = logs.Where(l => l.HasError == true).ToList();

            return output;
        }

        private ImportTimeAttendanceLog ParseLine(string lineContent, int lineNumber)
        {
            try
            {
                if (string.IsNullOrEmpty(lineContent))
                    // not considered error if it's just an empty line
                    // but it won't still be added to TimeAttendanceLog
                    return null;

                var parts = Regex.Split(lineContent, @"\t");

                if (parts.Length < 2)
                    return (new ImportTimeAttendanceLog()
                    {
                        LineContent = lineContent,
                        LineNumber = lineNumber,
                        ErrorMessage = "Needs at least 2 items in one line separated by a tab."
                    });

                if (string.IsNullOrEmpty(parts[0]) && string.IsNullOrEmpty(parts[1]))
                    // happens when blank lines was still read
                    // usually because it received an input like tab
                    return null;

                var employeeNo = parts[0].Trim();

                if ((Regex.Split(parts[1], " ").Count() < 2))
                    return (new ImportTimeAttendanceLog()
                    {
                        LineContent = lineContent,
                        LineNumber = lineNumber,
                        ErrorMessage = "Date log must consist of Date (1 space) Time."
                    });

                var logDate = ObjectUtils.ToNullableDateTime(parts[1]);

                if (logDate == null)
                    return (new ImportTimeAttendanceLog()
                    {
                        LineContent = lineContent,
                        LineNumber = lineNumber,
                        ErrorMessage = "Second column must be a valid Date Time."
                    });

                return (new ImportTimeAttendanceLog()
                {
                    EmployeeNumber = employeeNo,
                    DateTime = Convert.ToDateTime(logDate),
                    LineContent = lineContent,
                    LineNumber = lineNumber
                });
            }
            catch (Exception)
            {
                return (new ImportTimeAttendanceLog()
                {
                    LineContent = lineContent,
                    LineNumber = lineNumber,
                    ErrorMessage = "Error reading the line. Please check the template."
                });
            }
        }

        public class ImportOutput
        {
            public IList<ImportTimeAttendanceLog> Logs { get; set; }
            public IList<ImportTimeAttendanceLog> Errors { get; set; }

            /// <summary>
            /// True if the file was read successfully. Even if there are errors parsing
            /// some lines, as long as the file was read, this is still True.
            /// Posible reason for this to become False is when it did not find the chosen
            /// file.
            /// </summary>
            public bool IsImportSuccess { get; set; }

            public string ErrorMessage { get; set; }

            public ImportOutput()
            {
                this.Logs = new List<ImportTimeAttendanceLog>();
                this.Errors = new List<ImportTimeAttendanceLog>();
                this.IsImportSuccess = true;

                this.ErrorMessage = null;
            }
        }
    }

    public class ImportTimeAttendanceLog
    {
        // - Contents and error message
        public string LineContent { get; set; }

        public int LineNumber { get; set; }
        public string ErrorMessage { get; set; }

        // - Extracted contents
        public string EmployeeNumber { get; set; }

        public DateTime DateTime { get; set; }

        // - analyzed data
        public bool? IsTimeIn { get; set; }

        public DateTime? LogDate { get; set; }
        public Employee Employee { get; set; }
        public ShiftSchedule ShiftSchedule { get; set; }
        public EmployeeDutySchedule EmployeeDutySchedule { get; set; }
        public DateTime ShiftTimeInBounds { get; set; }
        public DateTime ShiftTimeOutBounds { get; set; }
        public string WarningMessage { get; set; }

        public ImportTimeAttendanceLog()
        {
            this.IsTimeIn = null;
            this.LogDate = null;

            this.EmployeeNumber = null;
        }

        public override string ToString() => DateTime.ToString("MM/dd/yyyy hh:mm tt");

        public string Type => IsTimeIn == null ? "" : (IsTimeIn.Value ? "Time-in" : "Time-out");

        public string EmployeeFullName => Employee == null ? "" : Employee.FullNameWithMiddleInitial;

        public bool HasError => ErrorMessage != null;

        public bool HasWarning => WarningMessage != null;

        public string ShiftDescription
        {
            get
            {
                if (ShiftSchedule != null && ShiftSchedule.Shift != null)
                    return $"{ShiftSchedule.Shift.TimeFrom.ToStringFormat("hh:mm tt")} - {ShiftSchedule.Shift.TimeTo.ToStringFormat("hh:mm tt")}";
                else if (EmployeeDutySchedule != null)
                    return $"{EmployeeDutySchedule.StartTime.ToStringFormat("hh:mm tt")} - {EmployeeDutySchedule.EndTime.ToStringFormat("hh:mm tt")}";

                return "-";
            }
        }

        public static List<IGrouping<string, ImportTimeAttendanceLog>> GroupByEmployee(IList<ImportTimeAttendanceLog> logs)
        {
            return logs.GroupBy(l => l.EmployeeNumber).ToList();
        }
    }
}